zcash_primitives/src/memo.rs

```diff
@@ -0,0 +1,411 @@
+//! Structs for handling encrypted memos.
+
+use std::error;
+use std::fmt;
+use std::ops::Deref;
+use std::str;
+use std::{cmp::Ordering, convert::TryFrom};
+
+/// Format a byte array as a colon-delimited hex string.
+///
+/// Source: https://github.com/tendermint/signatory
+/// License: MIT / Apache 2.0
+fn fmt_colon_delimited_hex<B>(f: &mut fmt::Formatter<'_>, bytes: B) -> fmt::Result
+where
+    B: AsRef<[u8]>,
+{
+    let len = bytes.as_ref().len();
+
+    for (i, byte) in bytes.as_ref().iter().enumerate() {
+        write!(f, "{:02x}", byte)?;
+
+        if i != len - 1 {
+            write!(f, ":")?;
+        }
+    }
+
+    Ok(())
+}
+
+/// Errors that may result from attempting to construct an invalid memo.
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    InvalidUtf8(std::str::Utf8Error),
+    TooLong(usize),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::InvalidUtf8(e) => write!(f, "Invalid UTF-8: {}", e),
+            Error::TooLong(n) => write!(f, "Memo length {} is larger than maximum of 512", n),
+        }
+    }
+}
+
+impl error::Error for Error {}
+
+/// The unencrypted memo bytes received alongside a shielded note in a Zcash transaction.
+#[derive(Clone)]
+pub struct MemoBytes(pub(crate) Box<[u8; 512]>);
+
+impl fmt::Debug for MemoBytes {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "MemoBytes(")?;
+        fmt_colon_delimited_hex(f, &self.0[..])?;
+        write!(f, ")")
+    }
+}
+
+impl Default for MemoBytes {
```

Comment:
I don't think that we should implement `Default` - instead, it's sufficient to just provide `MemoBytes::empty()`. I'm generally skeptical of `Default` as a trait, because adding `Default` as a constraint on a type passed to a method is almost never what you want to do, and unless you're using the trait as a constraint it's better to be explicit. We don't currently have any such `Default` constraints on methods, so I think we should remove this.

Reply:
Yeah, in this specific case I think an explicit `MemoBytes::empty` makes sense.

Reply:
However in the `Memo` case, I'm fine with `Default` existing, because unlike `MemoBytes` there's no chance of confusion between the "no memo" case and the empty byte slice case. The reason `Default` makes sense to implement is because it enables `#[derive(Default)]` to be used on downstream foreign types that include `Memo`. If those downstream types wanted to include `MemoBytes`, they would have to manually implement `Default` (now that we require an explicit `MemoBytes::empty()`).